### PR TITLE
Attach listeners via configuration

### PR DESCRIPTION
**Pull request by codeliner**
This is rough prototype of an EventManagerFactory which allows configuration of ListenerAggregates as well as listeners. This is just a proposal to continue discussion about it. It's neither tested nor complete!

### CODE REVIEW

File: src/Factory/EventManagerFactory.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+use Zend\ServiceManager\FactoryInterface;
+
+class EventManagerFactory implements FactoryInterface
+{
+    /**
+     * Override config key in an extending class to point to more specific key if needed
+     *
+     * @var string
+     */
+    protected $configKey = "event_manager";
```
Comment (bakura10) [inline src/Factory/EventManagerFactory.php]: Use simple quote for strings.

File: src/Factory/EventManagerFactory.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    public function __invoke(ServiceLocatorInterface $serviceLocator, $className, array $options = [])
+    {
+        // @TODO: ZF3 should have a Config accessible through FQCN
+        $config = $serviceLocator->get('Config');
+
+        $evm = new EventManager($serviceLocator->get(ListenerPluginManager::class));
```
Comment (bakura10) [inline src/Factory/EventManagerFactory.php]: $eventManager, no need for short names here :).

File: src/Factory/EventManagerFactory.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     * Override config key in an extending class to point to more specific key if needed
+     *
+     * @var string
+     */
+    protected $configKey = "event_manager";
+
+    /**
+     * {@inheritDoc}
+     */
+    public function __invoke(ServiceLocatorInterface $serviceLocator, $className, array $options = [])
```
Comment (bakura10) [inline src/Factory/EventManagerFactory.php]: I have renamed the $className to $requestedName in latest proposal of SM, as it's not 100% for sure a class name :).

File: src/Factory/EventManagerFactory.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        // @TODO: ZF3 should have a Config accessible through FQCN
+        $config = $serviceLocator->get('Config');
+
+        $evm = new EventManager($serviceLocator->get(ListenerPluginManager::class));
+
+        if (isset($config[$this->configKey])) {
```
Comment (bakura10) [inline src/Factory/EventManagerFactory.php]: Can be simplified to:

``` php
$config = isset($config[$this->configKey]['listener_aggregates'])
    ? $config[$this->configKey]['listener_aggregates']
    : [];

foreach ($config as $aggregateClass) {
 // ...
}
```

File: src/Factory/EventManagerFactory.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+             *     'my.event.name' => [
+             *        [[My\Listener1::class, 'onMyEventName'], 100],
+             *        [[My\Listener2::class, 'onMyEventName'], 100],
+             *     ]
+             *   ]
+             * ]
+             *
+             *
+             */
+            if (isset($config[$this->configKey]['event_listener_map'])) {
```
Comment (bakura10) [inline src/Factory/EventManagerFactory.php]: listener_aggregates is plural, but event_listener_map is singular? I think we could have better naming here :).

File: src/Factory/EventManagerFactory.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                foreach($config[$this->configKey]['event_listener_map'] as $eventName => $listeners) {
+                    foreach($listeners as $listenerSpecPriorityArr) {
```
Comment (bakura10) [inline src/Factory/EventManagerFactory.php]: you can use foreach ($listeners as list($listener, $spec))

**Comment by bakura10** [comment]
I actually kinda like the idea. Once the small fixes are done, I'll merge that to see what people think.

File: src/Factory/EventManagerFactory.php (hunk 5 above)
Comment (codeliner) [inline src/Factory/EventManagerFactory.php]: listener_aggregates is a list, whereby event_listener_map is a map :-) but yeah it's not looking good. What about simple `listeners` and `listener_aggregates`?

**Comment by codeliner** [comment]
:+1:

**Comment by bakura10** [comment]
We'll have a look at the naming after.

Could you please add a small example for listener aggregates too?

Please write some tests too and I'll merge that so other can discuss it.

Thanks :)

**Comment by codeliner** [comment]
@bakura10 Tests for the EventManagerFactory are a bit tricky due to the dependency to your refactor branch of the servicemanager. I got it working with composer but I don't think it makes sense to provide tests now as this is only a proposal and things will change.
But it is your choice. Maybe you have better idea to pull in the dependencies. My current version of the composer.json looks like that:

```
"repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/bakura10/zend-servicemanager.git"
        }
    ],
    "require-dev": {
        "fabpot/php-cs-fixer": "1.7.*",
        "satooshi/php-coveralls": "dev-master",
        "phpunit/PHPUnit": "~4.4",
        "zendframework/zend-servicemanager": "dev-refactor as 2.4.2"
    }
```

What do you think?

**Comment by bakura10** [comment]
Yep I think that's ok for now :).

File: src/Factory/EventManagerFactory.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+     */
+    protected $configKey = 'event_manager';
+
+    /**
+     * {@inheritDoc}
+     */
+    public function __invoke(ServiceLocatorInterface $serviceLocator, $requestedName, array $options = [])
+    {
+        // @TODO: ZF3 should have a Config accessible through FQCN
+        $config = $serviceLocator->get('Config');
```
Comment (bakura10) [inline src/Factory/EventManagerFactory.php:30]: You can do this:

``` php
$config = $serviceLocator->get('Config');
$config = isset($config[$this->configKey]) ? $config[$this->configKey] : [];
```

And then reuse $config without having to dereference with $this->configKey again and again.

**Comment by codeliner** [comment]
:+1:

File: src/Factory/EventManagerFactory.php (hunk 7 above)
Comment (codeliner) [inline src/Factory/EventManagerFactory.php:30]: I like to see the config structure in place, but I can change that.

**Comment by bakura10** [comment]
I'm going to merge, let's handle the test after.

File: src/Factory/EventManagerFactory.php [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        $config = isset($config[$this->configKey]) ? $config[$this->configKey] : [];
+
+        $eventManager = new EventManager($serviceLocator->get(ListenerPluginManager::class));
```
Comment (bakura10) [inline src/Factory/EventManagerFactory.php:33]: Actually this is wrong but I'll fix that tomorrow. The instantiator is not a plugin manager. The instantiator is just a callable that accepts a name.

**Comment by codeliner** [comment]
thanks for merging

File: src/Factory/EventManagerFactory.php (hunk 8 above)
Comment (codeliner) [inline src/Factory/EventManagerFactory.php:33]: ok, so the instantiator wraps a ListenerPluginManager or what is the purpose of the ListenerPluginManager?

**Pull request merged**